fdbserver/Ratekeeper.actor.cpp

```diff
@@ -853,23 +858,86 @@ ACTOR Future<Void> monitorThrottlingChanges(RatekeeperData *self) {
 	}
 }
 
-void tryAutoThrottleTag(RatekeeperData *self, StorageQueueInfo const& ss) {
-	if(ss.busiestTag.present() && ss.busiestTagFractionalBusyness > SERVER_KNOBS->AUTO_THROTTLE_TARGET_TAG_BUSYNESS && ss.busiestTagRate > SERVER_KNOBS->MIN_TAG_COST) {
-		TEST(true); // Transaction tag auto-throttled
+Future<Void> refreshStorageServerCommitCost(RatekeeperData *self) {
+	if(self->lastBusiestCommitTagPick == 0) { // the first call should be skipped
+		self->lastBusiestCommitTagPick = now();
+		return Void();
+	}
+	double elapsed = now() - self->lastBusiestCommitTagPick;
+	// for each SS, select the busiest commit tag from ssTrTagCommitCost
+	for(auto it = self->storageQueueInfo.begin(); it != self->storageQueueInfo.end(); ++it) {
+		it->value.busiestWriteTag.reset();
+		TransactionTag busiestTag;
+		TransactionCommitCostEstimation maxCost;
+		double maxRate = 0, maxBusyness = 0;
+		for(const auto& [tag, cost] : it->value.tagCostEst) {
+			// costBeforeSample / #sampledOps = COMMIT_SAMPLE_COST / 1
+			double rate = CLIENT_KNOBS->COMMIT_SAMPLE_COST * cost.getOpsSum() / elapsed;
```

Comment:
Since we have scaled cost, I think you might want to use something like `cost.getCostSum() / elapsed`. In the case that all of your ops are small, this is equivalent to what you have already, but in the case that your ops are big I think this would be more accurate. For example, if you have a single 150 cost op in 1 second, then `cost.getCostSum() / elapsed` would be 150/s, while I think `COMMIT_SAMPLE_COST * cost.getOpsSum() / elapsed` would be 100/s.

Reply:
Changed